### PR TITLE
Impersonation plugin

### LINKED ISSUE #6: Add from address impersonation plugin
| Q | A |
| --- | --- |
| Bug? | no |
| New Feature? | yes |

Add a new `ImpersonatePlugin` which really modifies the from address of the message sent.
#### Use case

When sending from staging/test systems within you network, but all the from address settings are already configured to the project domain (e.g. because the DB is synced from production every while), then those messages are not delivered due SPF and spam detection.

Thus the from address domain must be a address from the staging system is in control (SPF records are correct, etc.)

The `sender_address` of the SwiftmailerBundle is not sufficient for this case, as it modifies the `Return-Path` header which determines an address where bounce message and co should go. This is more a setting for production.
#### Additional informations

It's really similar to the `Swift_Plugins_ImpersonatePlugin` from Swiftmailer itself but modifying the `From` header.

The configuration in the bundle should be a simple parameter `from_address`.

### PR DESCRIPTION
**Pull request by angelov**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Fixed tickets | fixes #6 |
| Related issues/PRs | - |
| License | MIT |

This PR provides an option to change the sender's address before an email message is sent. The option is activated (for all or for any of the mailers separately) by adding the 'from_address' key in the configuration.

**Example:**

``` yaml
pixelart_swiftmailer_manipulator:
    from_address: 'impersonated@example.com'
```

### CODE REVIEW

**Comment by QuingKhaos** [comment]
Hi @angelov, many thanks for the pull request :tada: :tada: :tada: :tada: 

I will take a look later in the evening, may you already fix the failing travis test? Yes, sadly I still need to support PHP 5.4 :sob:

File: Swift/Plugins/ImpersonatePlugin.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ * (c) pixelart GmbH
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Pixelart\Bundle\SwiftmailerManipulatorBundle\Swift\Plugins;
+
+use Swift_Events_SendEvent;
```
Comment (QuingKhaos) [inline Swift/Plugins/ImpersonatePlugin.php]: Please remove the import and use `\Swift_Events_SendEvent` in the code below.

File: Swift/Plugins/ImpersonatePlugin.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    private $fromAddress;
+
+    /**
+     * @var string[]
+     */
+    private $originalAddresses = [];
+
+    /**
+     * @param string $address Address the message should be sent from
```
Comment (QuingKhaos) [inline Swift/Plugins/ImpersonatePlugin.php]: phpDoc parameter name is not consistent with the parameter :)

File: Swift/Plugins/ImpersonatePlugin.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    /**
+     * @param Swift_Events_SendEvent $event
+     */
+    public function beforeSendPerformed(Swift_Events_SendEvent $event)
+    {
+        $message = $event->getMessage();
+
+        $this->originalAddresses = $message->getFrom();
```
Comment (QuingKhaos) [inline Swift/Plugins/ImpersonatePlugin.php]: May you set the original address as an `X-Swift-From` into the message (headers), like the Swiftmailer does with [sender address impersonate](https://github.com/swiftmailer/swiftmailer/blob/a6d30977565081bc9b177b627bff2e2ba905224c/lib/classes/Swift/Plugins/ImpersonatePlugin.php#L46)?

File: Swift/Plugins/ImpersonatePlugin.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        $message->setFrom($this->fromAddress);
+    }
+
+    /**
+     * @param Swift_Events_SendEvent $event
+     */
+    public function sendPerformed(Swift_Events_SendEvent $event)
+    {
+        $message = $event->getMessage();
+        $message->setFrom($this->originalAddresses);
```
Comment (QuingKhaos) [inline Swift/Plugins/ImpersonatePlugin.php]: And restore the original from address from the mentioned `X-Swift-From` header above?

File: Tests/DependencyInjection/Fixtures/config/yml/full.yml [hunk 5]
```diff
@@ -1,3 +1,4 @@
 pixelart_swiftmailer_manipulator:
     prepend_subject: '[TESTSYSTEM!]'
     prepend_body: 'swiftmailer/prepend_body.txt.twig'
+    from_address: 'fake@example.com'
```
Comment (patkar) [inline Tests/DependencyInjection/Fixtures/config/yml/full.yml]: Missing newline on the end ;)

File: Tests/DependencyInjection/Fixtures/config/yml/one_mailer.yml [hunk 6]
```diff
@@ -3,3 +3,4 @@ pixelart_swiftmailer_manipulator:
         main_mailer:
             prepend_subject: '[TESTSYSTEM!]'
             prepend_body: 'swiftmailer/prepend_body.txt.twig'
+            from_address: 'fake@example.com'
```
Comment (patkar) [inline Tests/DependencyInjection/Fixtures/config/yml/one_mailer.yml]: Missing newline on the end ;)

File: DependencyInjection/PixelartSwiftmailerManipulatorExtension.php [hunk 7]
```diff
@@ -82,4 +83,38 @@ private function configureMailerManipulator($name, array $mailer, ContainerBuild
             ;
         }
     }
+
+    /**
+     * @param string           $name      The mailer name
+     * @param array            $mailer    The mailer configuration
+     * @param ContainerBuilder $container The container builder
+     */
+    private function configureMailerImpersonation($name, array $mailer, ContainerBuilder $container)
```
Comment (QuingKhaos) [inline DependencyInjection/PixelartSwiftmailerManipulatorExtension.php]: I would call it `configureMailerFromAddress`, because that's the config name we are using and configuring on the mailer. :grin:

**Review by QuingKhaos — CHANGES_REQUESTED** [review]

File: Tests/Swift/Plugins/ImpersonatePluginTest.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * @author Dejan Angelov <angelovdejan92@gmail.com>
+ */
+class ImpersonatePluginTest extends \PHPUnit_Framework_TestCase
+{
+    public function testImpersonatesSender()
+    {
+        $message = $this->prophesize('\Swift_Mime_Message');
+        $oldAddress = 'old@example.com';
```
Comment (QuingKhaos) [inline Tests/Swift/Plugins/ImpersonatePluginTest.php:24]: May you add a second test which tests with an array of multiple from addresses as `$oldAddress`. Not required for the merge, just more an edge case (:

**Comment by angelov** [comment]
@patkar I think you can re-check it now :)

**Review by QuingKhaos — APPROVED** [review]

**Comment by QuingKhaos** [comment]
Thanks for the pull request :tada: :tada: 

And released as [1.1.0](https://github.com/pixelart/swiftmailer-manipulator-bundle/releases/tag/1.1.0) :grinning:
